cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerModificationService.java

```diff
@@ -274,6 +286,91 @@ public void upgradeClusterRuntime(Set<ClusterComponent> components, boolean patc
         }
     }
 
+    private void tagHostsWithHostTemplateName() throws ApiException, CloudbreakException {
+        ClouderaManagerRepo clouderaManagerRepoDetails = clusterComponentProvider.getClouderaManagerRepoDetails(stack.getCluster().getId());
+        if (!isVersionNewerOrEqualThanLimited(clouderaManagerRepoDetails::getVersion, CLOUDERAMANAGER_VERSION_7_6_0)) {
+            LOGGER.info("Skipping host tagging. Cloudera Manager version needs to be equal or higher, than 7.6.0.");
+            return;
+        }
+        LOGGER.info("Tagging hosts after runtime upgrade.");
+        ApiClient v46Client = buildv46ApiClient();
+        ClustersResourceApi clustersResourceApi = clouderaManagerApiFactory.getClustersResourceApi(v46Client);
+        HostsResourceApi hostsResourceApi = clouderaManagerApiFactory.getHostsResourceApi(v46Client);
+        String clusterName = stack.getName();
+        Set<String> hostnamesFromCM = getClusterHostnamesFromCM(clustersResourceApi, clusterName);
+        Set<Call> callSet = Optional.of(stack).map(Stack::getInstanceGroups).stream().flatMap(Set::stream)
+                .map(asyncTagHostsInHostGroup(hostsResourceApi, hostnamesFromCM))
+                .flatMap(Set::stream)
+                .collect(Collectors.toSet());
+
+        callSet.forEach(call -> {
+            try {
+                if (!call.isExecuted()) {
+                    call.execute();
+                }
+            } catch (IOException e) {
+                LOGGER.info("Error while executing tagging.", e);
+                throw new ClouderaManagerOperationFailedException(e.getMessage(), e);
+            }
+        });
+    }
+
+    private ApiClient buildv46ApiClient() throws CloudbreakException {
+        ApiClient v46Client;
+        Cluster cluster = stack.getCluster();
+        String user = cluster.getCloudbreakAmbariUser();
+        String password = cluster.getCloudbreakAmbariPassword();
+        try {
+            v46Client = clouderaManagerApiClientProvider.getV46Client(stack.getGatewayPort(), user, password, clientConfig);
+        } catch (ClouderaManagerClientInitException e) {
+            LOGGER.error("Failed to init V46 client.", e);
+            throw new CloudbreakException(e);
+        }
+        return v46Client;
+    }
+
+    private Function<InstanceGroup, Set<Call>> asyncTagHostsInHostGroup(HostsResourceApi hostsResourceApi, Set<String> hostnamesFromCM) {
+        return instanceGroup -> Optional.ofNullable(instanceGroup).map(InstanceGroup::getNotDeletedInstanceMetaDataSet)
+                .stream().flatMap(Set::stream)
+                .map(InstanceMetaData::getDiscoveryFQDN)
+                .filter(hostnamesFromCM::contains)
+                .map(asyncTagHost(hostsResourceApi, instanceGroup))
+                .collect(Collectors.toSet());
+    }
+
+    private Function<String, Call> asyncTagHost(HostsResourceApi hostsResourceApi, InstanceGroup instanceGroup) {
+        return hostname -> {
```

Comment:
this method could be simplified a bit if it would return with a `Call` instead of a `Function`